### PR TITLE
feat(rust): apply left side predicate pushdown also to right side if all predicate columns are also join columns

### LINKED ISSUE #7247: Propagate predicate pushdown on left/inner join table
### Problem description

When I am filtering on a column `A` of a dataframe and then joining (left/inner) on this column on another dataframe. I wish polars could be able to apply the filter at scan level for the second dataframe also.

Is it possible ? On the roadmap ?

Ex:
```
import tempfile
import polars as pl
with tempfile.TemporaryDirectory() as temp_dir:
    file_path_a = Path(temp_dir) / "df_a.csv"
    file_path_b = Path(temp_dir) / "df_b.csv"
    
    
    pl.DataFrame({'A' : [0, 1, 2, 3, 4]}).write_parquet(file_path_a)
    pl.DataFrame({'A' : range(1000), 'B': 0}).write_parquet(file_path_b)
    
    df_a = pl.scan_parquet(file_path_a)
    df_b = pl.scan_parquet(file_path_b)
    
    r = df_a.filter(pl.col('A') == 0).join(df_b, on = ['A'], how='inner') # or left
    print(r.describe_optimized_plan())

#  INNER JOIN:
#  LEFT PLAN ON: [col("A")]
#    PARQUET SCAN /tmp/tmpgbghjv0b/df_a.csv
#    PROJECT */1 COLUMNS
#    SELECTION: [(col("A")) == (0i64)]
#  RIGHT PLAN ON: [col("A")]
#    PARQUET SCAN /tmp/tmpgbghjv0b/df_b.csv
#    PROJECT */2 COLUMNS
#    SELECTION: [(col("A")) == (0i64)]      # <- This does not exist today
#  END INNER JOIN

```

### PR DESCRIPTION
**Pull request by drgif**
Closes #7247

I have limited it to left and inner joins for the sole reason that I don't really have experience with the other types. If this can be extended, it is an easy fix.

### CODE REVIEW

File: crates/polars/tests/it/lazy/predicate_queries.rs [hunk 1]
```diff
@@ -269,3 +269,33 @@ fn test_predicate_pushdown_block_8847() -> PolarsResult<()> {
...
+    let out = df
+        .filter(col("a").eq(lit("a1")))
+        .filter(col("a").eq(col("c")))
+        .collect()?;
+    let expected = df![
+        "a" => ["a1"],
+        "b" => ["b1"],
+        "c" => ["a1"],
+    ]?;
+    assert_eq!(out, expected);
```
Comment (ritchie46) [inline crates/polars/tests/it/lazy/predicate_queries.rs:299]: This tests if the result is correct, but not if the optimization is applied to the right side. Can you test that?

Comment (ritchie46) [reply crates/polars/tests/it/lazy/predicate_queries.rs:299]: You you can iter the logical plans and assert all scan nodes have a predicate.

Maybe you can add an argument to this function: https://github.com/pola-rs/polars/blob/696a1f1ec784cd5e90903259c58e9f0c30f7d574/crates/polars-lazy/src/tests/optimization_checks.rs#L22

Where the argument influence if we do an `all` or an `any` check.

Comment (ritchie46) [reply crates/polars/tests/it/lazy/predicate_queries.rs:299]: Or maybe check if all tests still succeed if you change it to `all` and go for that.

**Comment by drgif** [comment]
Sorry I was a bit slow... Thanks for adding the test!

**Pull request merged**